WASM Verification by Redundancy

mplebanski opened the pull request:
Implements https://github.com/golemfactory/golem/issues/4421

kubkon commented on apps/glambda/glambdaenvironment.py:
```diff
@@ -26,5 +26,8 @@ def get_container_config(self) -> Dict:
+            # gVisor uses HOME variable before starting the image.
+            # Starting docker as a particular user (docker --user parameter)
+            # does not set HOME and that's why we define it here.
+            environment={'HOME': '/home/user'} if is_linux() else {}
```
How did `glambda` specifics creep in here? 😬

mplebanski replied:
Well I rebased to b0.20 and that was conflicted with our branch.

kubkon replied:
Ah, ok. We can fix that later then.

mbenke commented on apps/wasm/task.py:
```diff
@@ -154,6 +268,151 @@ def filter_task_results(self, task_results, subtask_id, log_ext=".log",
+            return 0.0
+
+        num_finished = len(list(filter(lambda x: x.is_finished(), self.subtasks)))
+        logger.info("progress: %f", num_finished / num_total)
```
this will probably log incorrect progress

kubkon replied:
Good catch!

mbenke commented on apps/wasm/task.py:
```diff
@@ -166,7 +425,7 @@ def build_full_definition(cls, task_type: 'CoreTaskTypeInfo',
         dictionary['options']['output_path'] = ''
         # Subtasks count is determined by the amount of subtask info provided.
-        dictionary['subtasks_count'] = len(dictionary['options']['subtasks'])
+        dictionary['subtasks_count'] = (WasmTask.REDUNDANCY_FACTOR + 1) * len(dictionary['options']['subtasks'])
```
This needs reconsideration

kubkon replied:
Agreed! We need to agree whether we want this to be transparent wrt the redundancy; i.e., whether Golem should or shouldn't be aware of verification by redundancy at all.

mbenke commented on apps/wasm/vbr.py:
```diff
@@ -0,0 +1,311 @@
+            actor {Actor} -- Actor who has computed the result
+            result {Any} --  Computation result
+        Raises:
+            UnknownActorError - raised when caller deliver an actor that was not previously reported by `add_actor` call.
```
may also raise `ValueError`

mplebanski replied:
Yeah I see that API description is not aligned to implementation comments. I will remove impl comments and update API.

mbenke commented on apps/wasm/vbr.py:
```diff
@@ -0,0 +1,311 @@
+        """
+        Returns:
+            Optional[List[Any, Actor, VerificationResult]] -- If verification is resolved a list of 3-element
+            tuples (result reference, actor, verification_result) is returned. A None is returned
```
```suggestion
            tuples (actor, result reference, verification_result) is returned. A None is returned
```

mbenke commented on apps/wasm/vbr.py:
```diff
@@ -0,0 +1,311 @@
+
+class BucketVerifier(VerificationByRedundancy):
+    """Simple verification by redundancy: subtask is executed by 2 providers,
+with a possible third if no decision can be reached based on the first 2."""
```
This comment is copied from another class and not really accurate

mbenke commented on apps/wasm/vbr.py:
```diff
@@ -0,0 +1,311 @@
+            if not found:
+                self.buckets.append(Bucket(self.comparator, key=result, value=actor))
+
+        print(f'add_actor: {len(self.actors)} actors, {len(self.results)} results, {len(self.buckets)} buckets')
```
This debug message should go away.

```suggestion
```

mbenke commented on apps/wasm/vbr.py:
```diff
@@ -0,0 +1,311 @@
+            self.more_actors_needed = False
+
+
+class SimpleSubtaskVerifier(VerificationByRedundancy):
```
This class is not used (and implements a different API anyway - int comparator rather than bool comparator`).
I would remove it (or replace with correct version).

mplebanski replied:
Removed.

mplebanski commented on apps/wasm/task.py:
```diff
@@ -1,20 +1,85 @@
+        return self.subtasks.keys()
+
+    def add_result(self, s_id, task_result):
+        self.verifier.add_result(self.subtasks[s_id]["actor"], task_result)
```
I believe crashing hard is what we actually want in this case. Giving a wrong actor or a result for not reported actor is not what we intend.

mplebanski commented on apps/wasm/task.py:
```diff
@@ -1,20 +1,85 @@
+
+    def get_verdicts(self):
+        verdicts = []
+        for actor, result, verdict in self.verifier.get_verdicts():
```
`VbrSubtask` has method `is_finished` to make sure it never happens.

mbenke replied:
..and if `is_finished ` is not checked, we crash with a `TypeError`

kubkon replied:
IMHO any redundant checks in `get_verdicts` are not strictly required. All we need is to document the use-case in the docs, as in "call `is_finished` before calling `get_verdicts`".

mplebanski commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+        return None
+
+    def _cmp_results(result_list_a: List[Any], result_list_b: List[Any]) -> bool:
+        for r1, r2 in zip(result_list_a, result_list_b):
```
I don't know. @kubkon ?

kubkon replied:
AFAIK, yep, they ought to be unless something went horribly wrong in one Golem node; e.g., computation failed, etc.

mplebanski commented:
@mbenke I modified `add_actor` but then reverted the changes. There was a problem that two instances of an Actor class with the same uuid were not equal (which is expected, but we forgot about it), thus duplicated actor was not detected. I added __eq__ and __hash__ operations to Actor class.

mbenke commented on apps/wasm/vbr.py:
```diff
@@ -0,0 +1,181 @@
+        return self.uuid == other.uuid
+
+    def __hash__(self):
+        return id(self)
```
Why not 
```suggestion
        return hash(self.uuid)
```

kubkon commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+                b1 = f1.read()
+                b2 = f2.read()
+                if b1 != b2:
+                    logger.info("Verification of task failed")
```
It's a remnant of the old code. But I agree, we should iron it out. Also, it probably should be `debug` log rather than `info`.

kubkon commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+                    return False
+
+        logger.info("Comparing: %s and %s", result_list_a, result_list_b)
+        logger.info("Verification of task was successful")
```
See above.

kubkon commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+
+    def computation_finished(self, subtask_id, task_result,
+                             verification_finished=None) -> None:
+        logger.info("Called in WasmTask")
```
Again, see above. I've used it to trace calls through Golem. Not as easy as it looks :grin:

mplebanski replied:
I removed it. There are other more informative logs left though.

kubkon commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+
+                if verdict == VerificationResult.SUCCESS:
+                    # pay up!
+                    logger.info("Accepting results for subtask %s", s_id)
```
That's a good idea. We should probably also change from `info` to `debug`.

kubkon commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+                    TaskClient.assert_exists(actor.uuid, self.counting_nodes).accept()
+                else:
+                    logger.info("Rejecting results for subtask %s", s_id)
+                    self.subtasks_given[s_id]['status'] = SubtaskStatus.failure
```
I'm not sure I understand your question. Could you elaborate?

kubkon commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+
+        verdicts = subtask.get_verdicts()
+
+        s_ids = [s_id for s_id in subtask.get_instances()]
```
Perhaps, although is that strictly necessary here? IHMO, it's debatable whether it improves the code hygiene here even.

kubkon commented on apps/wasm/task.py:
```diff
@@ -85,51 +139,111 @@ def __init__(self, total_tasks: int, task_definition: WasmTaskDefinition,
+        s_ids = [s_id for s_id in subtask.get_instances()]
+        for s_id in s_ids:
+            instance = subtask.get_instance(s_id)
+            for actor, verdict in verdicts:
```
We do? I'm not sure I can see that here. Could you explain?

mplebanski commented on apps/wasm/task.py:
```diff
@@ -1,20 +1,85 @@
+logger = logging.getLogger("apps.wasm")
+
+
+class VbrSubtask:
```
Resolved

mplebanski commented:
@mbenke @kubkon Please approve for control build.

maaktweluit commented:
@mplebanski it has an approval so it should build, strange.. hit the rebuild to check https://buildbot.golem.network/buildbot/#/builders/12/builds/10804

kubkon commented:
Closing as it's been superseded by #4475.
